### PR TITLE
Add Hive connector procedure call to create empty partition

### PR DESCRIPTION
**Pull request by wenleix**

### CODE REVIEW

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 1]
```diff
@@ -871,6 +871,32 @@ public void testCastNullToColumnTypes()
...
+            getQueryRunner().execute(getSession(), sql);
+            assertQuery(
+                    "SELECT count(*) FROM \"test_insert_partitioned_bucketed_table$partitions\"",
+                    "SELECT " + (i + 1));
+
+            assertQueryFails(sql, "Partition already exists.*");
+        }
+
+        assertUpdate("DROP TABLE " + tableName);
+        assertFalse(getQueryRunner().tableExists(getSession(), tableName));
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: We don't need this check here, as the purpose of this test is not to validate `DROP TABLE` behavior.

File: presto-orc/src/main/java/com/facebook/presto/orc/stream/LongOutputStreamV2.java [hunk 2]
```diff
@@ -117,7 +117,6 @@ public void writeLong(long value)
             initializeLiterals(value);
             return;
         }
-
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/stream/LongOutputStreamV2.java]: Unrelated change

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 3]
```diff
@@ -871,6 +871,32 @@ public void testCastNullToColumnTypes()
...
+    {
+        testCreateEmptyBucketedPartition(HiveStorageFormat.RCBINARY);
+    }
+
+    public void testCreateEmptyBucketedPartition(HiveStorageFormat storageFormat)
+    {
+        String tableName = "test_insert_partitioned_bucketed_table";
+        createPartitionedBucketedTable(tableName, storageFormat);
+
+        ImmutableList<String> orderStatusList = ImmutableList.of("F", "O", "P");
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Declare as `List`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveConnectorFactory.java [hunk 4]
```diff
@@ -124,6 +129,7 @@ public Connector create(String connectorId, Map<String, String> config, Connecto
             HiveSessionProperties hiveSessionProperties = injector.getInstance(HiveSessionProperties.class);
             HiveTableProperties hiveTableProperties = injector.getInstance(HiveTableProperties.class);
             ConnectorAccessControl accessControl = new PartitionsAwareAccessControl(injector.getInstance(ConnectorAccessControl.class));
+            Set<Procedure> procedures = injector.getInstance(Key.get(new TypeLiteral<Set<Procedure>>(){}));
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveConnectorFactory.java]: Please run the code formatter. I think there is supposed to be a space here: `() {}`

(search the code base for other usages of `TypeLiteral` to confirm)

File: presto-hive/src/main/java/com/facebook/presto/hive/CreateEmptyPartitionProcedure.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    {
+        HiveMetadata hiveMetadata = hiveMetadataFactory.create();
+
+        HiveInsertTableHandle hiveInsertTableHandle = hiveMetadata.beginInsert(session, new HiveTableHandle(schema, table));
+
+        List<String> actualPartitionColumnNames = hiveInsertTableHandle.getInputColumns().stream()
+                .filter(HiveColumnHandle::isPartitionKey)
+                .map(HiveColumnHandle::getName)
+                .collect(toImmutableList());
+        checkArgument(Objects.equals(partitionColumnNames, actualPartitionColumnNames), "input partition column names doesn't match actual partition column names");
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/CreateEmptyPartitionProcedure.java]: I think this needs to be a `PrestoException(INVALID_PROCEDURE_ARGUMENT)` as it means they are passing in invalid partition names. Add a test for this.

File: presto-hive/src/main/java/com/facebook/presto/hive/CreateEmptyPartitionProcedure.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+import static com.facebook.presto.spi.block.MethodHandleUtil.methodHandle;
+import static com.facebook.presto.spi.type.StandardTypes.VARCHAR;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class CreateEmptyPartitionProcedure
+        implements Provider<Procedure>
+{
+    private static final MethodHandle INSERT_EMPTY_PARTITION = methodHandle(
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/CreateEmptyPartitionProcedure.java]: Let's all this `CREATE_EMPTY_PARTITION` to be consistent with the class and method name

File: presto-hive/src/main/java/com/facebook/presto/hive/CreateEmptyPartitionProcedure.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+            throw new PrestoException(ALREADY_EXISTS, "Partition already exists");
+        }
+        String partitionName = FileUtils.makePartName(actualPartitionColumnNames, partitionStringValues);
+
+        WriteInfo writeInfo = locationService.getPartitionWriteInfo(hiveInsertTableHandle.getLocationHandle(), Optional.empty(), partitionName);
+        hiveMetadata.finishInsert(
+                session,
+                hiveInsertTableHandle,
+                ImmutableList.of(
+                    Slices.wrappedBuffer(
```
[the remark this answers is not in the document]
Comment (wenleix) [reply presto-hive/src/main/java/com/facebook/presto/hive/CreateEmptyPartitionProcedure.java]: @electrum : Is this hard to read because the purpose of `Slices.wrappedBuffer` is unclear? So the variable name could be something like `serializedPartitionUpdate`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java (hunk 1 above)
Comment (wenleix) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: @electrum : The code is refactored from the original tests. I think it's used as a "post-condition" assertion to make sure the table is really cleaned up and to make sure different tests are well separated. 

I can see your point that the tests doesn't really check the `DROP TABLE`. And I am OK either way :) .
